### PR TITLE
Directly initialize properties in UserMetadata

### PR DESCRIPTION
**Pull request by rsgowman**
Prompted by attempting to enable tsconfig options
'strictPropertyInitialization'.

The properties were being indirectly set via Object.defineProperty.
Typescript isn't able to follow that. Setting the properties directly
resolves it. (And also exposed some minor type errors.)

### CODE REVIEW

File: src/auth/user-record.ts [hunk 1]
```diff
@@ -66,16 +66,19 @@ export interface CreateRequest extends UpdateRequest {
  * @constructor
  */
 export class UserMetadata {
-  public readonly creationTime: string;
-  public readonly lastSignInTime: string;
+  public readonly creationTime?: string | null;
```
Comment (rsgowman) [inline src/auth/user-record.ts]: The `?` isn't strictly necessary as `parseDate` always returns `null` if the input date is undefined. But I think that's more implementation detail-ish, so I've left open the possibility of undef here.

File: src/index.d.ts [hunk 2]
```diff
@@ -491,13 +491,12 @@ declare namespace admin.auth {
     /**
      * The date the user last signed in, formatted as a UTC string.
      */
-    lastSignInTime: string;
+    lastSignInTime?: string | null;
```
Comment (rsgowman) [inline src/index.d.ts]: nb: technically, a public api change. Arguably breaking since typscript users may have to adjust their code. (Note that we've *always* allowed this to be null... we've just not defined it as such. IOW, I think this is a bugfix more than a change.)

File: src/auth/user-record.ts (hunk 1 above)
Comment (hiranya911) [reply src/auth/user-record.ts]: These are not currently defined as optional (both here and in the d.ts file). So may be we can keep it as it was?

File: src/index.d.ts (hunk 2 above)
Comment (hiranya911) [reply src/index.d.ts]: I think the `?` is more of a significant change. Can we remove that?

File: src/auth/user-record.ts (hunk 1 above)
Comment (rsgowman) [reply src/auth/user-record.ts]: The problem is that these reflect the proto, so when the values come back from the server, there isn't any guarantee that these will actually be populated. (For instance, the lastSignInTime will not be present if the user has never logged in.) As it happens, our implementation will notice that and set lastSignInTime to null. I'm not sure if that happens for all of our objects or not... though we could explicitly set things to null in that situation should there be cases where things are otherwise left out, thereby dropping the '?'.

But not adding the '| null' isn't great since it could legitimately be null (or missing). We just didn't notice before because the property was being set indirectly. We could detect this and use a sentinel value instead (empty string in this case?). To be clear, in some circumstances, we are setting these non-nullable values to null today; typescript just isn't noticing.

In other words, I suspect many instances of this throughout the code base are legitimately optional and/or nullable, but because they've been set indirectly, we haven't noticed.

Comment (hiranya911) [reply src/auth/user-record.ts]: Marking is nullable is fine (and is the right thing to do, since we explicitly initialize it to null). But marking it optional (`?`) seems wrong to me, since it's never undefined.

Comment (rsgowman) [reply src/auth/user-record.ts]: Ah, I understand you now. I've removed the '?'.

Note that other classes (that I haven't made this conversion to yet) may not explicitly set to null and instead just not set the property. (Possibly; haven't checked yet.) I'll set the types to reflect what they are (like we've now done here) and we can figure out then if there's an inconsistency issue that needs addressing.

File: src/index.d.ts [hunk 3]
```diff
@@ -491,13 +491,12 @@ declare namespace admin.auth {
     /**
      * The date the user last signed in, formatted as a UTC string.
      */
-    lastSignInTime: string;
+    lastSignInTime: string | null;
```
Comment (hiranya911) [inline src/index.d.ts]: Nit: Let's also update the doc comment to indicate when these values are null.

Comment (rsgowman) [reply src/index.d.ts]: Done. Though note that there's not a great explanation as to why creationTime might be null. As an alternative, we could keep it as non-nullable and change the implementation to throw an internal error if this situation occurs.

Comment (hiranya911) [reply src/index.d.ts]: That sounds like the right approach to me.

Comment (rsgowman) [reply src/index.d.ts]: Done. A number of unit tests needed to be changed as a result. I think the changes are correct, but definitely will want Bassam to weigh in on this too.
